Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39920 Enable dynamic quantized GRU support
* **#37159 Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell**
* #38851 Refactor LSTM tests

Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

raghuramank10000 committed:
Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell

Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

jamesr66a commented:
Can you also add deprecation warnings to `torch/jit/quantized.py` in these functions: https://github.com/pytorch/pytorch/blob/9b3911c0737d1d12650cb668dc40bb979bc81eb8/torch/jit/quantized.py#L491

Once this patch lands we should move all users over to this IMO. It's getting to be an increasing maintenance burden to keep the `jit/quantized.py` code around. Thanks!

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -1320,6 +1319,7 @@ bool _use_cudnn_rnn_flatten_weight() {
         std::move(packed_output.data), std::move(std::get<1>(result)));     \
   }
 
+
```
nit: extra line

supriyar commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -1368,15 +1368,15 @@ std::tuple<Tensor, Tensor> quantized_gru_data_legacy(
-      dropout_p,
-      train,
-      bidirectional);
+                            data,
```
formatting seems off

supriyar commented on test/quantization/test_quantize.py:
```diff
@@ -677,51 +678,71 @@ def checkQuantized(model):
+                if torch.backends.quantized.engine == 'qnnpack' and dtype == torch.float16:
+                    continue                            # fp16 dynamic quant is not supported for qnnpack
+
+                print(dtype)
```
remove print statement?

raghuramank100 replied:
Sure, will clean this up prior to landing

supriyar commented on test/quantization/test_quantize.py:
```diff
@@ -761,12 +782,50 @@ def test_per_channel_lstm_quantize(self):
+                if torch.backends.quantized.engine == 'qnnpack' and dtype == torch.float16:
+                    continue                            # fp16 dynamic quant is not supported for qnnpack
+
+                print(dtype)
```
extra print here too

supriyar reviewed (approved):
lgtm, left a few comments

raghuramank100 commented on test/quantization/test_quantize.py:
```diff
@@ -677,51 +678,71 @@ def checkQuantized(model):
+
+            for dtype in [torch.qint8, torch.float16]:
+                if torch.backends.quantized.engine == 'qnnpack' and dtype == torch.float16:
+                    continue                            # fp16 dynamic quant is not supported for qnnpack
```
Not sure if it is off here?

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell"


Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

[ghstack-poisoned]
